pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterMetricsStats.java

```diff
@@ -0,0 +1,208 @@
+ *   the big record itself is not written in batch hence is not included in the batch metrics written above (batch
+ *   size, batch delay, etc). The trigger is of course counted as other trigger types.
+ */
+public class TxnLogBufferedWriterMetricsStats implements Closeable {
+
+    static final double[] RECORD_COUNT_PER_ENTRY_BUCKETS = {10, 50, 100, 200, 500, 1_000};
+
+    static final double[] BYTES_SIZE_PER_ENTRY_BUCKETS = {128, 512, 1_024, 2_048, 4_096, 16_384,
+            102_400, 1_232_896};
+
+    static final double[] MAX_DELAY_TIME_BUCKETS = {1, 5, 10};
+
+    @Getter
+    private final String metricsPrefix;
+
+    private final String[] labelNames;
+
+    private final String[] labelValues;
+
+    /** Count of records in per transaction log batch. **/
+    private final Histogram recordsPerBatchMetric;
+    private final Histogram.Child recordsPerBatchHistogram;
+
+    /** Bytes size per transaction log batch. **/
+    private final Histogram batchSizeBytesMetric;
+    private final Histogram.Child batchSizeBytesHistogram;
+
+    /** The time of the oldest transaction log spent in the buffer before being sent. **/
+    private final Histogram oldestRecordInBatchDelayTimeSecondsMetric;
+    private final Histogram.Child oldestRecordInBatchDelayTimeSecondsHistogram;
+
+    /** The count of the triggering transaction log batch flush actions by "batchedWriteMaxRecords". **/
+    private final Counter batchFlushTriggeredByMaxRecordsMetric;
+    private final Counter.Child batchFlushTriggeredByMaxRecordsCounter;
+
+    /** The count of the triggering transaction log batch flush actions by "batchedWriteMaxSize". **/
+    private final Counter batchFlushTriggeredByMaxSizeMetric;
+    private final Counter.Child batchFlushTriggeredByMaxSizeCounter;
+
+    /** The count of the triggering transaction log batch flush actions by "batchedWriteMaxDelayInMillis". **/
+    private final Counter batchFlushTriggeredByMaxDelayMetric;
+    private final Counter.Child batchFlushTriggeredByMaxDelayCounter;
+
+    /**
+     * If {@link TxnLogBufferedWriter#asyncAddData(Object, TxnLogBufferedWriter.AddDataCallback, Object)} accept a
+     * request that param-data is too large (larger than "batchedWriteMaxSize"), then two flushes are executed:
+     *    1. Write the data cached in the queue to BK.
+     *    2. Direct write the large data to BK.
+     * This ensures the sequential nature of multiple writes to BK.
+     */
+    private final Counter batchFlushTriggeredByLargeSingleDataMetric;
+    private final Counter.Child batchFlushTriggeredByLargeSingleDataCounter;
+
+    /**
+     * Users needs to ensure that the {@link TxnLogBufferedWriterMetricsStats} of the same {@param metricsPrefix} can
+     * only create once, otherwise an IllegalArgumentException will be thrown.
+     */
+    public TxnLogBufferedWriterMetricsStats(String metricsPrefix, String[] labelNames, String[] labelValues,
```

Comment:
Does this Class singleton?
If singleton, it's better to:
```
private static TxnLogBufferedWriterMetricsStats stats;
public static TxnLogBufferedWriterMetricsStats create(Args... args) {
    if (stats == null) {
         stats = new TxnLogBufferedWriterMetricsStats(args);
    }
    return stats;
}
``` 
if not a singleton, 
```
       recordsPerBatchMetric = new Histogram.Builder()
                        .name(recordsPerBatchMetricName)
                        .labelNames(labelNames)
                        .help("Records per batch histogram")
                        .buckets(RECORD_COUNT_PER_ENTRY_BUCKETS)
                        .register(registry);
```
may lead to a Exception if `registry` is always a same instance.

Reply:
@tjiuming This is what I consulted with you if you remember a couple of weeks ago, and based on this I wrote: https://github.com/apache/pulsar/pull/16758#issuecomment-1207782630

Reply:
Hi @tjiuming @asafm 

Yes, this has been discussed.

TxnLogBufferedWriterMetricsStats only two instances, they are holds by `MLTransactionMetadataStoreProvider` and `TransactionPendingAckStoreProvider`.  In the following PR, there will be changes like this: 

- Append modifier `abstract` to `TxnLogBufferedWriterMetricsStats`
- Create an internal class in `MLTransactionMetadataStoreProvider` and `TransactionPendingAckStoreProvider` like this:

```java
private static class MLTransactionMetadataStoreBufferedWriterMetrics extends TxnLogBufferedWriterMetricsStats{

        private MLTransactionMetadataStoreBufferedWriterMetrics() {
            super("pulsar_txn_tc", new String[]{"cluster", "broker"}, new String[]{"xxx", "xxx"}, CollectorRegistry.defaultRegistry);
        }
}
```

How are you feeling?

Reply:
@asafm @poorbarcode Oh sorry, I didn't noticed that a metricName is follow the pattern `%s_bufferedwriter_batch_record_count`, then it won't lead to an Exception.

Overall LGTM